### PR TITLE
Fix builder class generating name clash if union contains field named "error"

### PR DESCRIPTION
**Pull request by luqasn**
Unfortunately, there seems to be an upstream bug in addition to the missing fully qualified reference to kotlin.error in the code generator.

Or do you have an idea why my fix would not work?

### CODE REVIEW

**Comment by benjamin-bader** [comment]
kotlinpoet hasn't yet had a release containing your fix; are you testing with a local build to know that there's an issue?

It seems like there's an active _situation_ around Dokka and the move from jcenter to Maven Central, so CI isn't even running the tests you've added.  I'll see if I can clear that up.

**Comment by benjamin-bader** [comment]
OK, now CI shows your test-case failing.

As far as why your fix isn't working, my guess is that it's because `kotlin.error` is a function, not a type, and thus wouldn't be affected by the regular CodeWriter import rules.  Just a guess though.  Top-level functions are a rather late addition to kotlinpoet.

**Comment by luqasn** [comment]
Yeah, I think you are right. This seems to be a new, albeit similar upstream bug, using master does not fix it.

**Comment by luqasn** [comment]
> I think a better, if less-satisfying, fix will be to update KotlinCodeGenerator to reserve the name `error`. We'd do that in the function that generates `NameAllocator` instances, around line 155 or so. This will result in the field being named something like `error_` which kind of sucks, but will at least compile :)

Better as in quicker? Because to me, this is a clear KotlinPoet bug again. If I fully-qualify the reference I expect it to stay fully qualified if there is potential for a clash.

**Comment by benjamin-bader** [comment]
Better as in quicker, yes.  I share your expectation.

**Comment by luqasn** [comment]
There is no time pressure for the bug, so I would rather fix it upstream (unless extremely complicated or something). Will have a look what I can find.

**Comment by luqasn** [comment]
Not the prettiest solution: https://github.com/square/kotlinpoet/pull/1040

**Comment by luqasn** [comment]
Marked this as ready for merge because thrifty's side is done, now we are waiting for a new KotlinPoet version including my fix. An alternative to waiting until we can merge would be to merge the required code (the fully qualified method) and test in `@Disabled` state right now so we don't keep this open.

File: thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt
Comment (benjamin-bader) [inline thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt:360]: I'm so confused.  Why does a generated interface need to include the `abstract` keyword?  Isn't it implicit in the fact the the function belongs to an interface and has no body?

Comment (luqasn) [reply thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt:360]: Totally agree with you. KotlinPoet is a bit weird here. Although the redundant modifier is only cosmetic, I am curious and checking out the relevant code :)

Comment (benjamin-bader) [reply thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt:360]: I see; I guess it may be another "library-mode" change.  That sure junked up our generated code.  Guess it's a net win though.

Comment (luqasn) [reply thrifty-kotlin-codegen/src/test/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGeneratorTest.kt:360]: The weird thing is not that it now emits the keyword when you explicitly specify it (that sounds very much correct).
The weird thing is that you can not NOT explicitly make a function in an interface abstract (which is why thrifty also needs to add an abstract marker).
The notion of implicit abstractness of a interface function does not exist.

**Comment by benjamin-bader** [comment]
@luqasn looks like kotlinpoet 1.8.0 just shipped, including your recent bugfixes!

**Comment by luqasn** [comment]
Moved the updating to https://github.com/microsoft/thrifty/pull/446, will rebase after that's merged

**Comment by luqasn** [comment]
Ready to merge now. Thankfully, the KotlinPoet folks even took the very ugly "interface methods emit explicit abstract modifier" back :)

**Review by benjamin-bader — APPROVED** [review]

**Comment by benjamin-bader** [comment]
Thanks for seeing this through - I really appreciate how thorough you've been!

**Comment by luqasn** [comment]
> Thanks for seeing this through - I really appreciate how thorough you've been!

You are more than welcome! Thanks for being so responsive and supportive!

**Pull request merged**
